### PR TITLE
[libc++][format] Handle range-underlying-spec

### PR DESCRIPTION
**Pull request by poyaoc97**
An immediate colon signifeis that the range-format-spec contains only range-underlying-spec.

This patch allows this code to compile and run:
```c++
std::println("{::<<9?}", std::span<const char>{"Hello", sizeof "Hello"});
```

### CODE REVIEW

**Review by mordante — COMMENTED** [review]
Thanks for working on this. Since the patch is a draft I mainly glanced over it.

I'm curious, why did you add `__has_range_underlying_spec_` instead of using `__use_range_fill_`?

**Comment by poyaoc97** [comment]
Because the tuple formatter like the range formatter prohibits ':' as a `fill` (__use_range_fill_ = true) but it doesn't have an underlying format-spec. So `{::<}` for tuples is a no-go but OK (align left, no fill for the underlying formatter) for ranges (`__has_range_underlying_spec_ = true` for this case).

**Comment by mordante** [comment]
> Because the tuple formatter like the range formatter prohibits ':' as a `fill` (__use_range_fill_ = true) but it doesn't have an underlying format-spec. So `{::<}` for tuples is a no-go but OK (align left, no fill for the underlying formatter) for ranges (`__has_range_underlying_spec_ = true` for this case).

I expected that motivation. I think it would be more future proof to handle tuple the same and get a slightly worse error message. If tuple gets an underlying format-spec we're less likely to reintroduce this bug. (I am investigating this feature for a future C++ proposal.)

**Comment by poyaoc97** [comment]
> I think it would be more future proof to handle tuple the same and get a slightly worse error message.

OK, will update the patch tmr. Looks like it would bump into `std::__throw_format_error("The format specifier should consume the input or end with a '}'");`, no big deal.

> If tuple gets an underlying format-spec we're less likely to reintroduce this bug. (I am investigating this feature for a future C++ proposal.)

Do you have a timeline you would like to share? :)
Would be cool to see it on the C++26 train.

**Comment by mordante** [comment]
> > If tuple gets an underlying format-spec we're less likely to reintroduce this bug. (I am investigating this feature for a future C++ proposal.)
> 
> Do you have a timeline you would like to share? :) Would be cool to see it on the C++26 train.

My time-line heavily depends on the available time I have ;-) Still I hope to get a paper done before the end of next month.

**Comment by poyaoc97** [comment]
Ping~

File: libcxx/test/std/utilities/format/format.range/format.range.fmtset/format.functions.tests.h [hunk 1]
```diff
@@ -156,7 +157,7 @@ void test_char_string(TestFunction check, [[maybe_unused]] ExceptionTest check_e
 
   check_exception("The format string contains an invalid escape sequence", SV("{:}<s}"), input);
   check_exception("The fill option contains an invalid value", SV("{:{<s}"), input);
-  check_exception("The fill option contains an invalid value", SV("{::<s}"), input);
+  check_exception("The type option contains an invalid value for a character formatting argument", SV("{::<s}"), input);
```
Comment (mordante) [inline libcxx/test/std/utilities/format/format.range/format.range.fmtset/format.functions.tests.h:160]: Can you move this test to the "type" testing part (line 181 in the original code).

**Review by mordante — COMMENTED** [review]
> Ping~

Sorry I didn't have time to look at this earlier.

Mainly looks good one small nit.

File: libcxx/test/std/utilities/format/format.range/format.range.formatter/format.functions.tests.h [hunk 2]
```diff
@@ -157,7 +158,7 @@ void test_char_string(TestFunction check, ExceptionTest check_exception, auto&&
 
   check_exception("The format string contains an invalid escape sequence", SV("{:}<s}"), input);
   check_exception("The fill option contains an invalid value", SV("{:{<s}"), input);
-  check_exception("The fill option contains an invalid value", SV("{::<s}"), input);
+  check_exception("The type option contains an invalid value for a character formatting argument", SV("{::<s}"), input);
```
Comment (mordante) [inline libcxx/test/std/utilities/format/format.range/format.range.formatter/format.functions.tests.h:161]: Sorry I wasn't clear, but I meant that for this one too.

**Review by mordante — APPROVED** [review]
One more change and then I'm happy with the patch.

**Pull request merged**